braindecode/datautil/serialization.py

```diff
@@ -98,14 +101,107 @@ def _load_signals_and_description(path, preload, raws, ids_to_load=None):
     for i in ids_to_load:
         fif_file = os.path.join(path, file_name.format(i))
-        all_signals.append(_load_signals(fif_file, preload, raws))
+        all_signals.append(_load_signals(fif_file, preload, is_raw))
     description_df = description_df.iloc[ids_to_load]
     return all_signals, description_df
 
 
-def _load_signals(fif_file, preload, raws):
-    if raws:
+def _load_signals(fif_file, preload, is_raw):
+    if is_raw:
         signals = mne.io.read_raw_fif(fif_file, preload=preload)
     else:
         signals = mne.read_epochs(fif_file, preload=preload)
     return signals
+
+
+def load_concat_dataset(path, preload, ids_to_load=None, target_name=None,
+                        n_jobs=1):
+    """Load a stored BaseConcatDataset of BaseDatasets or WindowsDatasets from
+    files.
+
+    Parameters
+    ----------
+    path: str
+        Path to the directory of the .fif / -epo.fif and .json files.
+    preload: bool
+        Whether to preload the data.
+    ids_to_load: None | list(int)
+        Ids of specific files to load.
+    target_name: None or str
+        Load specific description column as target. If not given, take saved
+        target name.
+    n_jobs: int
+        Number of jobs to be used to read files in parallel.
+
+    Returns
+    -------
+    concat_dataset: BaseConcatDataset of BaseDatasets or WindowsDatasets
+    """
+    # if we encounter a dataset that was saved in 'the old way', call the
+    # corresponding 'old' loading function
+    if _is_outdated_saved(path):
+        warnings.warn("The way your dataset was saved is deprecated by now. "
+                      "Please save it again using dataset.save().", UserWarning)
+        return _outdated_load_concat_dataset(
+            path=path, preload=preload, ids_to_load=ids_to_load,
+            target_name=target_name)
+
+    # else we have a dataset saved in the new way with
+    # - subdirectories in path for every dataset
+    # - description.json and -epo.fif or -raw.fif in every subdirectory
+    # - target_name.json in path if we were given raws and target specified
+    # grab all the fif and json files, numbers have to match
+    fif_files = glob(os.path.join(path, '**/*.fif'))
+    assert len(fif_files) > 0, 'No fif files found.'
+    description_files = glob(os.path.join(path, '**/description.json'))
+    assert len(fif_files) == len(description_files), (
+        'Number of descriptions and .fif files does not match.')
+    # sort files by id of the datasets
+    fif_files = sorted(fif_files, key=lambda p: int(p.split(os.sep)[-2]))
+    description_files = sorted(description_files,
+                               key=lambda p: int(p.split(os.sep)[-2]))
+    # optionally make a selection of files
+    if ids_to_load is not None:
+        fif_files = [fif_files[i] for i in ids_to_load]
+        description_files = [description_files[i] for i in ids_to_load]
+    is_raw = fif_files[0].endswith('-raw.fif')
```

Comment:
maybe very edge case, but what would happen if we have both raw and epochs in one folder? just for my understanding

Reply:
If the raws are not fif, then everything should work as expected (would have to verify this).

If raws are stored as fif, it should not be possible to create this situation through braindecode functionality. If the user did this himself, then
For saving:
- our code will crash if overwrite = False
- our code will delete all -raw.fif and -epochs.fif and description.json with overwrite = True

For loading:
- will crash if number of .fif files and description.json files does not match
- it will crash later since the reading function is determined by first fif file, so either we will try to read raws with mne.read_epochs or epochs with mne.io.read_raw_fif


We could add an explicit test somewhere before this happens. So check that not -raw.fif and -epo.fif are in the directory as well as only allow -raw.fif / -epo.fif and description.json files. Crash if anything else exists? Or how to check for other raw data files?